pennylane/_grad.py

```diff
@@ -181,3 +182,206 @@ def _jacobian_function(*args, **kwargs):
+
+    Args:
+        f (function): function with signature ``f(*args, **kwargs)``
+        argnum (int): which argument to take a gradient with respect to
+        delta (float): step size used to evaluate the finite difference
+        idx (list[int]): if argument ``args[argnum]`` is an array, ``idx`` can
+            be used specify the indices of the arguments to differentiate
+
+    Returns:
+        (float or array): the gradient of the input function with respect
+        to the arguments in ``argnum``
+    """
+
+    if argnum > len(args) - 1:
+        raise ValueError(
+            "The value of 'argnum' has to be between 0 and {}; got {}".format(len(args) - 1, argnum)
+        )
+
+    x = _np.array(args[argnum])
+    gradient = _np.zeros_like(x, dtype="O")
+
+    if x.ndim == 0 and idx is not None:
+        raise ValueError(
+            "Argument {} is not an array, 'idx' should be set to 'None';"
+            " got {}".format(argnum, idx)
+        )
+
+    if idx is None:
+        idx = list(_np.ndindex(*x.shape))
+    else:
+        bounds = _np.array(x.shape) - 1
+        for _idx in idx:
+
+            if (_np.array(_idx) > bounds).any():
+                raise ValueError(
+                    "Indices in 'idx' can not be greater than {}; got {}".format(
+                        tuple(bounds), _idx
+                    )
+                )
+
+    for i in idx:
+        shift = _np.zeros_like(x)
+        shift[i] += 0.5 * delta
+        gradient[i] = (
+            f(*args[:argnum], x + shift, *args[argnum + 1 :], **kwargs)
+            - f(*args[:argnum], x - shift, *args[argnum + 1 :], **kwargs)
+        ) * delta ** -1
+
+    return gradient
+
+
+def _fd_second_order_centered(f, argnum, delta, *args, idx=None, **kwargs):
+    r"""Uses a central finite difference approximation to compute the second-order
+    derivative :math:`\frac{\partial^2 f(x)}{\partial x_i \partial x_j}` of the function ``f``
+    with respect to the argument ``argnum``.
+
+    Args:
+        f (function): function with signature ``f(*args, **kwargs)``
+        argnum (int): which argument to take the derivative with respect to
+        delta (float): step size used to evaluate the finite differences
+        idx (list[int]): if argument ``args[argnum]`` is an array, `idx`` specifies
+            the indices ``i, j`` of the arguments to differentiate
+
+    Returns:
+        (float or array): the second-order derivative of the input function with respect
+        to the arguments in ``argnum``
+    """
+
+    if argnum > len(args) - 1:
+        raise ValueError(
+            "The value of 'argnum' has to be between 0 and {}; got {}".format(len(args) - 1, argnum)
+        )
+
+    x = _np.array(args[argnum])
+
+    if x.ndim == 0 and idx is not None:
+        raise ValueError(
+            "Argument {} is not an array, 'idx' should be set to 'None';"
+            " got {}".format(argnum, idx)
+        )
+
+    if idx is None:
+        idx = [(), ()]
```

Comment:
Yep, this confused me as well. I _think_ this might be sneakily using a feature of NumPy that `array[tuple()]` simply returns the whole array.

Reply:
Yep, this seems to be the case. 
![image](https://user-images.githubusercontent.com/2959003/111303330-5f2a8700-868f-11eb-80eb-93e827931568.png)

I didn't know this @agran2018, you taught me something 😆 

Definitely worth commenting this within the code though, so that it is clear what is happening.

Reply:
I think there is some confusion here. This function gives the second derivative of the Hamiltonian. We don't need the full Hessian of the Hamiltonian as they enter individually in the calculation of the second derivative of the energy. Then,

- For multivariable functions ``idx`` is a list with the indices ``i`` and ``j`` involved in the second derivative.
- for one variable function ``idx`` defaults to ``None`` and by having ``idx= [(), ()]`` we can compute the second derivative of one variable functions using the code block implementing the diagonal case.

Indeed @josh146, I learned this from the first function you proposed :-).